aopy/preproc/laser.py

```diff
@@ -0,0 +1,171 @@
+        warnings.warn(f"{np.count_nonzero(missing_times)} unmeasured laser offsets")
+        corrected_off_times[missing_off_times] = laser_event_times[missing_off_times] + laser_event_widths[missing_off_times]
+    times_not_found = np.logical_or(np.isnan(corrected_idx), np.isnan(corrected_off_idx))
+
+    # Now calculate the widths and powers based on corrected times
+    laser_sensor_widths = corrected_off_times - corrected_times
+    laser_on_times = np.mean([corrected_off_times, corrected_times], axis=0)
+    laser_on_samples = (laser_on_times * ds_fs).astype(int)
+    laser_sensor_volts = (ds_data[laser_on_samples])*sensor_voltsperbit
+    laser_sensor_powers = calibrate_sensor(laser_sensor_volts, peak_power_mW)  
+
+    # Correct the widths and powers with the given thresholds
+    laser_event_powers = calibrate_gain(laser_event_gains, peak_power_mW, calibration_file=calibration_file)
+    corrected_widths, widths_above_thr = base.validate_measurements(laser_event_widths, laser_sensor_widths, thr_width)
+    corrected_powers, powers_above_thr = base.validate_measurements(laser_event_powers, laser_sensor_powers, thr_power)
+
+    if debug:
+        print(f"BMI3D recorded {len(laser_event_times)} stims")
+        print(f"Laser sensor crossed threshold {len(laser_sensor_times)} times")
+        if len(laser_event_times) == len(laser_sensor_times):
+            print(f"Average difference: {np.mean(laser_event_times - laser_sensor_times):.4f} s")
+        else:
+            print("Cannot compute average difference. Check the laser threshold is correct")
+            print(f"Using threshold: {thr_volts}")
+            print(f"Maximum voltage deviation of sensor: {np.max(ds_data)*sensor_voltsperbit}")
+
+        plt.figure()
+        visualization.plot_laser_sensor_alignment(ds_data*sensor_voltsperbit, ds_fs, corrected_times)
+
+    return corrected_times, corrected_widths, corrected_powers, times_not_found, widths_above_thr, powers_above_thr
+
+def calibrate_gain(gain, peak_power_mW, calibration_file='qwalor_447nm_ch2.yaml'):
+    '''
+    Convert gain into laser power in mW
+
+    Args:
+        gain ((ntrial,) float array): sensor readings in volts
+        peak_power_mW (float): peak power in mW
+        calibration_file (str, optional): name of calibration file. Defaults to 'qwalor_447nm_ch2.yaml'.
+
+    Returns:
+        (ntrial,) float array: laser power in mW
+
+    Examples:
+
+        .. code-block:: python
+
+            gain = np.arange(0.4,1.0,0.01)
+            powers = preproc.laser.calibrate_laser_gain(gain, 12.)
+            plt.figure()
+            plt.plot(gain, powers)
+            plt.xlabel('gain')
+            plt.ylabel('power (mW)')
+
+        .. image:: _images/calibrate_laser_gain.png
+    '''
+    config_dir = files('aopy').joinpath('config')
+    with as_file(config_dir.joinpath(calibration_file)) as f:
+        yaml_data = aodata.yaml_read(f)
+    voltages = np.interp(gain, yaml_data['gain'], yaml_data['power'])
```

Comment:
This is probably fine for now but since the relationship between power and gain is nonlinear we should probably fit an exponential to the measurements and then calculate the power from that.